src/Renci.SshNet/Session.cs

```diff
@@ -669,6 +672,129 @@ public void Connect()
+                // ToDo: SHould we really limit this globally here, or leave it to the app?
+                AuthenticationConnection.Wait();
+
+                if (IsConnected)
+                    return;
+
+                //lock (this)   // ToDo: Why do we need double locking at all?
+                {
+                    // If connected don't connect again
+                    if (IsConnected)
+                        return;
+
+                    // Reset connection specific information
+                    Reset();
+
+                    // Build list of available messages while connecting
+                    _sshMessageFactory = new SshMessageFactory();
+
+                    _socket = await _serviceFactory.CreateConnector(ConnectionInfo, _socketFactory)
+                                             .ConnectAsync(ConnectionInfo, cancellationToken).ConfigureAwait(false);
+
+                    var serverIdentification = await _serviceFactory.CreateProtocolVersionExchange()
+                                                              .StartAsync(ClientVersion, _socket, cancellationToken).ConfigureAwait(false);
+
+                    // Set connection versions
+                    ServerVersion = ConnectionInfo.ServerVersion = serverIdentification.ToString();
+                    ConnectionInfo.ClientVersion = ClientVersion;
+
+                    DiagnosticAbstraction.Log(string.Format("Server version '{0}' on '{1}'.", serverIdentification.ProtocolVersion, serverIdentification.SoftwareVersion));
+
+                    if (!(serverIdentification.ProtocolVersion.Equals("2.0") || serverIdentification.ProtocolVersion.Equals("1.99")))
+                    {
+                        throw new SshConnectionException(string.Format(CultureInfo.CurrentCulture, "Server version '{0}' is not supported.", serverIdentification.ProtocolVersion),
+                                                         DisconnectReason.ProtocolVersionNotSupported);
+                    }
+
+                    // Register Transport response messages
+                    RegisterMessage("SSH_MSG_DISCONNECT");
+                    RegisterMessage("SSH_MSG_IGNORE");
+                    RegisterMessage("SSH_MSG_UNIMPLEMENTED");
+                    RegisterMessage("SSH_MSG_DEBUG");
+                    RegisterMessage("SSH_MSG_SERVICE_ACCEPT");
+                    RegisterMessage("SSH_MSG_KEXINIT");
+                    RegisterMessage("SSH_MSG_NEWKEYS");
+
+                    // Some server implementations might sent this message first, prior to establishing encryption algorithm
+                    RegisterMessage("SSH_MSG_USERAUTH_BANNER");
+
+                    // Mark the message listener threads as started
+                    _messageListenerCompleted.Reset();
+
+                    // Start incoming request listener
+                    // ToDo: Make message pump async, to not consume a thread for every session
```

Comment:
Are you planning on working on this?

Reply:
Yes, this is something that we should definitely do, but I'd rather leave this for a later PR. This PR adds a new client surface area and I think we should first finish this as I think this is stable. Async message pump on the other hand is an implementation detail that can be changed later. And we need to be clear on the compilation targets that we want to support, because I have a feeling that supporting both sync (for pre-net40 targets) and async message pump in the codebase will be a royal PITA.